File: backend/src/InsuraTech.Application/Policies/Commands/CreatePolicy/Strategies/CreateTravelPolicyStrategy.cs
using InsuraTech.Application.Common.Interfaces;
using InsuraTech.Domain.Policies;
using InsuraTech.Domain.Policies.TravelPlan;
using InsuraTech.Domain.Policies.ValueObjects;

namespace InsuraTech.Application.Policies.Commands.CreatePolicy.Strategies;

public sealed class CreateTravelPolicyStrategy : ICreatePolicyStrategy
{
    private readonly ITrmService _trmService;

    public CreateTravelPolicyStrategy(ITrmService trmService) => _trmService = trmService;

    public PolicyType Type => PolicyType.Travel;

    public bool CanHandle(CreatePolicyCommand request) =>
        request.TripType.HasValue && request.DurationDays.HasValue;

    public async Task<Policy> CreateAsync(CreatePolicyCommand request, PolicyNumber policyNumber, InsuredPerson insured, CancellationToken cancellationToken)
    {
        // Para Travel se usa el constructor directo para evitar la validación de mínimo 30 días
        // (un viaje de 1-29 días es válido en Travel aunque no lo sea en otros productos)
        var coverage = new CoveragePeriod(request.CoverageStartDate, request.CoverageEndDate);

        decimal? trmValue = null;
        DateOnly? trmDate = null;

        if (request.TripType == TripType.Internacional)
        {
            var trm  = await _trmService.GetCurrentTrmAsync(cancellationToken);
            trmValue = trm.ValueCop;
            trmDate  = trm.Date;
        }

        return Policy.CreateTravelPolicy(
            policyNumber, insured, coverage,
            request.TripType!.Value, request.Continent,
            request.DurationDays!.Value, trmValue, trmDate);
    }
}

File: backend_v2/src/main/java/com/insuratech/application/policies/strategies/CreateTravelPolicyStrategy.java
// Spec: SPEC-014 — travel-plan
// Origen: CreateTravelPolicyStrategy.cs → CreateTravelPolicyStrategy.java
package com.insuratech.application.policies.strategies;

import com.insuratech.application.policies.commands.CreatePolicyCommand;
import com.insuratech.domain.policies.Policy;
import com.insuratech.domain.policies.PolicyType;
import com.insuratech.domain.policies.travel.TravelPlanSelection;
import com.insuratech.domain.policies.vo.CoveragePeriod;
import com.insuratech.domain.policies.vo.InsuredPerson;
import com.insuratech.domain.policies.vo.PolicyNumber;
import org.springframework.stereotype.Component;

@Component
public class CreateTravelPolicyStrategy implements PolicyCreationStrategy {

    @Override
    public PolicyType type() { return PolicyType.TRAVEL; }

    @Override
    public boolean canHandle(CreatePolicyCommand cmd) {
        return cmd.tripType() != null;
    }

    @Override
    public Policy create(CreatePolicyCommand cmd, PolicyNumber policyNumber, InsuredPerson insured) {
        int travelers = cmd.numberOfTravelers() != null ? cmd.numberOfTravelers() : 1;
        var selection = new TravelPlanSelection(
            cmd.tripType(), cmd.travelDestinations(),
            cmd.coverageStartDate(), cmd.coverageEndDate(), travelers);
        var period = CoveragePeriod.of(cmd.coverageStartDate(), cmd.coverageEndDate());
        return Policy.createTravelPolicy(policyNumber, insured, period, selection, cmd.agentId(), cmd.cityId());
    }
}
